### PR TITLE
Add tests for the behavior of unsubscribing affecting next dispatch a…

### PR DESCRIPTION
**Pull request by slorber**
…nd not current dispatch listeners (decision of gaearon in #1180)

### CODE REVIEW

File: test/createStore.spec.js [hunk 1]
```diff
@@ -286,6 +286,66 @@ describe('createStore', () => {
...
+    const listener4 = expect.createSpy(() => {})
+    unsubscribeHandles.push(store.subscribe(listener4))
+    const listener5 = expect.createSpy(() => {})
+    unsubscribeHandles.push(store.subscribe(listener5))
+
+    store.dispatch(unknownAction())
+    store.dispatch(unknownAction())
+    expect(listener1.calls.length).toBe(1)
+    expect(listener2.calls.length).toBe(1)
+    // listener3 is the one removing the others so obviously it got executed
```
Comment (gaearon) [inline test/createStore.spec.js]: I'd still prefer an explicit check to an “obvious” comment ;-)

File: test/createStore.spec.js [hunk 2]
```diff
@@ -286,6 +286,66 @@ describe('createStore', () => {
...
+    expect(listener4.calls.length).toBe(1)
+    expect(listener5.calls.length).toBe(1)
+  })
+
+  it('not fire immediately if a listener is added inside another listener', () => {
```
Comment (gaearon) [inline test/createStore.spec.js]: Nit: `it('does not`

File: test/createStore.spec.js [hunk 3]
```diff
@@ -286,6 +286,66 @@ describe('createStore', () => {
...
+    const store = createStore(reducers.todos)
+
+    const listener1 = expect.createSpy(() => {})
+    const listener2 = expect.createSpy(() => {})
+    const listener3 = expect.createSpy(() => {})
+
+    let listener3Added = false
+    const maybeAddThirdListener = () => {
+      if ( !listener3Added ) {
```
Comment (gaearon) [inline test/createStore.spec.js]: Nit: we don't put spaces around conditions. (Why doesn't ESLint catch that?)

File: test/createStore.spec.js [hunk 4]
```diff
@@ -286,6 +286,66 @@ describe('createStore', () => {
...
+        listener3Added = true
+        store.subscribe(() => listener3())
+      }
+    }
+
+    // Listener 1 is normal
+    store.subscribe(() => listener1())
+
+    // Listener 2 adds the 3rd listener another listener
```
Comment (gaearon) [inline test/createStore.spec.js]: I think we can remove these comments, the code is pretty descriptive

**Comment by gaearon** [comment]
Thanks! This is a valuable PR. Always good to ensure we don't break promises.
I'll be happy to merge after you address a few nits.

File: test/createStore.spec.js (hunk 3 above)
Comment (slorber) [inline test/createStore.spec.js]: oooh ok i hate this rule :)

Comment (gaearon) [inline test/createStore.spec.js]: FWIW I hate no-semicolons but other folks from `rackt` use this style. Consistency :100:

**Comment by slorber** [comment]
@gaearon I've updated the code it should be fine now

np that will be my first merged PR on redux ecosystem :)

**Comment by gaearon** [comment]
Appreciated. Congrats!

**Pull request merged**